### PR TITLE
Add g3dinference element

### PR DESCRIPTION
**Pull request by ZiningLi**
### Description

This PR introduces a new 3D inference element, g3dinference, for running PointPillars-based 3D object detection on LiDAR point clouds in DL Streamer. The new element consumes application/x-lidar buffers produced by g3dlidarparse, validates and reuses LiDAR metadata, executes the PointPillars voxelization, neural network, and post-processing stages with OpenVINO, and attaches flattened 3D detection tensors for downstream conversion and publishing.

In addition to the new element itself, this PR extends the existing 3D pipeline integration by updating LiDAR metadata, wiring the new element into the 3D plugin build and registration flow, **adding JSON conversion support** for 3D detection results, and ensuring metadata publishing works correctly for LiDAR pipelines. It also **adds documentations**, **PointPillars model download support**, **sample scripts**, and **end-to-end unit test** coverage for the new pipeline.

Fixes  [ITEP-88267](https://jira.devtools.intel.com/browse/ITEP-88267)

### What's New
**New 3D Element: `g3dinference`**
- Adds a new 3D GStreamer element, g3dinference, for PointPillars-based LiDAR inference.
- Consumes application/x-lidar buffers with LidarMeta from g3dlidarparse.
- Runs a 3-stage PointPillars pipeline: voxelization, NN inference, and post-processing.
- Loads runtime settings from a JSON config with voxel_params, extension_lib, voxel_model, nn_model, and postproc_model.
- Supports config, device, model-type, and score-threshold properties.

**3D Detection Metadata**
- Attaches **GstGVATensorMeta** with flattened 3D detection output in [N, 9] format: x, y, z, w, l, h, theta, confidence, label_id
- Marks output as pointpillars_3d for downstream conversion and publishing.

**LidarMeta Update**
- Extends LidarMeta with exit_g3dinference_timestamp to improve timing visibility across the LiDAR pipeline.

**Docs, PointPillars model download support, sample and test**

- PointPillars model download support: 

image
image

- sample: 

image
image
image

- test:

image
image

### Changed Files Overview
```text
docs/user-guide/
└── elements/
    ├── elements.md # Add g3dinference reference to the 3D elements overview
    └── g3dinference.md # New g3dinference element documentation

include/dlstreamer/gst/metadata/
└── g3d_lidar_meta.h # Extend LidarMeta with g3dinference exit timestamp

samples/
├── download_public_models.sh # Add pointpillars model download support
└── gstreamer/
    └── gst_launch/
        └── g3dinference/
            ├── README.md # Sample usage and prerequisites
            ├── g3dinference.sh # Sample PointPillars inference pipeline
            └── g3dinference_prepare.sh # Prepare models, config, and extension library

src/gst/metadata/
└── g3d_lidar_meta.c # Update LidarMeta implementation for new timestamp field

src/monolithic/gst/
├── 3d_elements/
│   ├── CMakeLists.txt # Wire g3dinference build
│   └── g3dinference/
│       ├── CMakeLists.txt # Element build
│       ├── g3dinference.cpp # PointPillars inference and tensor metadata logic
│       └── g3dinference.h # Element header and property definitions
├── 3d_registrator/
│   ├── CMakeLists.txt # Register/link 3D elements
│   └── register_3d_elements.cpp # Plugin registration for g3dinference
└── elements/
    ├── gvametaconvert/
    │   ├── gstgvametaconvert.c # Integration updates for metadata conversion flow
    │   └── jsonconverter.cpp # Add JSON serialization for PointPillars 3D output
    └── gvametapublish/
        └── base/
            └── gvametapublishbase.cpp # Handle LiDAR buffers correctly in publish flow

tests/unit_tests/tests_gstgva/
├── run_tests.py # Register g3dinference test module
└── test_pipeline_g3dinference.py # End-to-end PointPillars pipeline test
```
### Any Newly Introduced Dependencies

No new mandatory third-party build dependencies were introduced into the core DL Streamer build.

For samples and tests, this PR relies on existing PointPillars assets from openvino_contrib and requires the OpenVINO Python package to build the PointPillars custom extension library used by the sample and unit test flow.

### How Has This Been Tested?

Can follow the sample pipeline under `samples/gstreamer/gst_launch/g3dinference`.

Manual:
```bash
gst-launch-1.0 multifilesrc location="lidar/%06d.bin" caps=application/octet-stream ! \
  g3dlidarparse ! \
  g3dinference config=pointpillars_ov_config.json device=GPU score-threshold=0.5 ! \
  gvametaconvert add-tensor-data=true format=json json-indent=2 ! \  
  gvametapublish file-format=2 file-path=pointpillar.json ! \
  fakesink
```
### Checklist:

- [x] I agree to use the MIT license for my code changes.
- [x] I have not introduced any 3rd party components incompatible with MIT. 
- [x] I have not included any company confidential information, trade secret, password or security token. 
- [x] I have performed a self-review of my code.

### CODE REVIEW

**Review by yunowo — DISMISSED** [review]

File: tests/unit_tests/tests_gstgva/test_pipeline_g3dinference.py [hunk 1]
```diff
@@ -0,0 +1,322 @@
...
+        for field_name in ("x", "y", "z", "w", "l", "h", "theta"):
+            self.assertAlmostEqual(
+                actual["bbox_3d"][field_name], expected["bbox_3d"][field_name], delta=self.BOX_DELTA,
+                msg=f"bbox_3d.{field_name} mismatch for object {index}"
+            )
+
+    def _assert_detections_match(self, payload):
+        actual_detections = self._extract_detected_objects(payload)
+        self._print_all_detections_debug(actual_detections)
+        if len(actual_detections) != len(EXPECTED_DETECTIONS):
```
[the remark this answers is not in the document]
Comment (ZiningLi) [reply tests/unit_tests/tests_gstgva/test_pipeline_g3dinference.py:459]: The full detection dump was added to make the bbox validation path easier to inspect when this test is run in debug mode, since this test verifies not only detection count but also per-object bbox values, confidence, and labels. In normal runs, this debug output is not shown unless the test is executed with higher verbosity such as -vv, so it should not affect standard CI log readability.

File: src/monolithic/gst/3d_elements/g3dinference/g3dinference.cpp [hunk 2]
```diff
@@ -0,0 +1,514 @@
...
+            if (path.is_absolute())
+                return path.lexically_normal().string();
+            return (config_dir / path).lexically_normal().string();
+        };
+
+        _extension_lib = resolve_path(config_json.at("extension_lib").get<std::string>());
+        _voxel_model_path = resolve_path(config_json.at("voxel_model").get<std::string>());
+        _nn_model_path = resolve_path(config_json.at("nn_model").get<std::string>());
+        _postproc_model_path = resolve_path(config_json.at("postproc_model").get<std::string>());
+        _device = device;
```
[the remark this answers is not in the document]
Comment (ZiningLi) [reply src/monolithic/gst/3d_elements/g3dinference/g3dinference.cpp:105]: voxel_params is used at PointPillars model export time rather than being applied separately by g3dinference at runtime. The generated XML artifacts already encode these voxelization settings, so the runtime only needs to load the exported models and extension library. I updated the documentation to make that distinction explicit.

File: docs/user-guide/elements/g3dinference.md [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+
+The element adds `GstGVATensorMeta` with:
+
+- `element_id = g3dinference`
+- `model_name = pointpillars` or the configured model type
+- `layer_name = pointpillars_3d_detection`
+- `format = pointpillars_3d`
+- `dims = [N, 9]`
+
+Each detection row contains 9 FP32 values. The first dimension `N` is the number of detections in the frame, and the second dimension is fixed at 9 values per detection:
```
[the remark this answers is not in the document]
Comment (ZiningLi) [reply docs/user-guide/elements/g3dinference.md]: The detection tensor payload here is stored as float data, so reporting GVA_PRECISION_FP16 was inconsistent. I updated the metadata to GVA_PRECISION_FP32 and synced the documentation to match the actual buffer format.

File: src/monolithic/gst/3d_elements/g3dinference/g3dinference.cpp [hunk 4]
```diff
@@ -0,0 +1,514 @@
...
+                     GST_TIME_ARGS(timestamp));
+    return timestamp;
+}
+
+void set_tensor_metadata(GstGVATensorMeta *tensor_meta, const std::vector<float> &detections, const char *model_type) {
+    gst_structure_set_name(tensor_meta->data, "detection");
+    gst_structure_set(tensor_meta->data, "element_id", G_TYPE_STRING, "g3dinference", "model_name", G_TYPE_STRING,
+                      model_type ? model_type : DEFAULT_MODEL_TYPE, "layer_name", G_TYPE_STRING,
+                      "pointpillars_3d_detection", "format", G_TYPE_STRING, "pointpillars_3d", "precision", G_TYPE_INT,
+                      GVA_PRECISION_FP16, "layout", G_TYPE_INT, GVA_LAYOUT_NC, "rank", G_TYPE_INT, 2, NULL);
```
[the remark this answers is not in the document]
Comment (ZiningLi) [reply src/monolithic/gst/3d_elements/g3dinference/g3dinference.cpp]: The detection tensor payload here is stored as float data, so reporting GVA_PRECISION_FP16 was inconsistent. I updated the metadata to GVA_PRECISION_FP32 and synced the documentation to match the actual buffer format.

**Review by yunowo — DISMISSED** [review]

File: docs/user-guide/elements/g3dinference.md [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+
+```json
+{
+  "voxel_params": {
+    "voxel_size": [0.16, 0.16, 4],
+    "point_cloud_range": [0, -39.68, -3, 69.12, 39.68, 1],
+    "max_num_points": 32,
+    "max_voxels": 16000
+  },
+  "extension_lib": "/home/kpi/pointPillars/ov_extensions/build/libov_pointpillars_extensions.so",
```
Comment (oonyshch) [inline docs/user-guide/elements/g3dinference.md]: please replace `/home/kpi/pointPillars/...` with generic placeholders like `/path/to/model/` 
as these appear to be dev machine paths

File: src/monolithic/gst/3d_elements/g3dinference/g3dinference.cpp [hunk 6]
```diff
@@ -0,0 +1,531 @@
...
+
+    g_object_class_install_property(gobject_class, PROP_MODEL_TYPE,
+                                    g_param_spec_string("model-type", "Model Type", "3D detector model type",
+                                                        DEFAULT_MODEL_TYPE,
+                                                        (GParamFlags)(G_PARAM_READWRITE | G_PARAM_STATIC_STRINGS)));
+
+    g_object_class_install_property(
+        gobject_class, PROP_SCORE_THRESHOLD,
+        g_param_spec_float("score-threshold", "Score Threshold",
+                           "Drop detections below this score (-1 keeps postproc output unchanged)", -1.0, 1.0,
```
Comment (oonyshch) [inline src/monolithic/gst/3d_elements/g3dinference/g3dinference.cpp]: please consider using `0.0` as default with documentation that `0.0` means no filtering, or add a `score-threshold-enabled` boolean property. using `-1` as a sentinel value is less clear.

File: docs/user-guide/elements/g3dinference.md [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+| model-type | String | 3D detector model type. Currently only `pointpillars` is supported. | pointpillars |
+| score-threshold | Float | Drops detections below this score. `-1` keeps all post-processing output unchanged. | -1.0 |
+
+## Configuration
+
+The `config` property should point to a PointPillars JSON configuration file. In practice, the file contains the paths to the OpenVINO extension library and the three models used by the runtime. It may also include `voxel_params` for compatibility with the exported PointPillars config format.
+
+Expected top-level entries:
+
+- `voxel_params`: PointPillars voxelization settings such as `voxel_size`, `point_cloud_range`, `max_num_points`, and `max_voxels`. These values are used when exporting the PointPillars models and are already encoded in the generated voxelization model; `g3dinference` does not currently apply them as runtime-tunable parameters.
```
Comment (oonyshch) [inline docs/user-guide/elements/g3dinference.md:32]: this note is easy to miss. please consider adding a warning block in the markdown (like `> **Warning:** ...`) and emitting a `GST_WARNING` if the config json contains `voxel_params`, so users know these are ignored at runtime

File: src/monolithic/gst/elements/gvametaconvert/jsonconverter.cpp [hunk 8]
```diff
@@ -499,6 +501,114 @@ json convert_radar_process_meta(GstGvaMetaConvert *converter, GstBuffer *buffer)
...
+        {"point_count", lidar_meta->lidar_point_count},
+        {"exit_lidarparse_timestamp", lidar_meta->exit_lidarparse_timestamp},
+        {"exit_g3dinference_timestamp", lidar_meta->exit_g3dinference_timestamp},
+    });
+
+    return result;
+}
+
+bool is_pointpillars_detection_tensor(const GVA::Tensor &tensor) {
+    if (tensor.format() != "pointpillars_3d")
```
Comment (oonyshch) [inline src/monolithic/gst/elements/gvametaconvert/jsonconverter.cpp:539]: these changes affect existing elements used across all pipelines. please add dedicated tests for the `pointpillars_3d` branch and verify no regression on existing tensor/detection output formats.

File: tests/unit_tests/tests_gstgva/test_pipeline_g3dinference.py [hunk 9]
```diff
@@ -0,0 +1,331 @@
...
+        self.config_file = os.path.join(self.test_dir, "pointpillars_ov_config.json")
+        self.extension_lib = self.__class__.extension_lib
+        self.voxel_model = os.path.join(self.pointpillars_root, "pretrained", "pointpillars_ov_pillar_layer.xml")
+        self.nn_model = os.path.join(self.pointpillars_root, "pretrained", "pointpillars_ov_nn.xml")
+        self.postproc_model = os.path.join(self.pointpillars_root, "pretrained", "pointpillars_ov_postproc.xml")
+
+    def tearDown(self):
+        shutil.rmtree(self.test_dir, ignore_errors=True)
+
+    def _write_runtime_config(self):
```
Comment (oonyshch) [inline tests/unit_tests/tests_gstgva/test_pipeline_g3dinference.py:184]: please add test cases for: (1) empty point cloud input, (2) mismatched `lidar_point_count` vs actual buffer size, (3) missing `config` property. these edge cases need explicit coverage.

**Review by oonyshch — CHANGES_REQUESTED** [review]
please take a look at the comments

File: docs/user-guide/elements/g3dinference.md (hunk 5 above)
Comment (ZiningLi) [reply docs/user-guide/elements/g3dinference.md]: Changed in the latest code.
  "extension_lib": "/path/to/pointPillars/ov_extensions/build/libov_pointpillars_extensions.so",
  "voxel_model": "/path/to/pointPillars/pretrained/pointpillars_ov_pillar_layer.xml",
  "nn_model": "/path/to/pointPillars/pretrained/pointpillars_ov_nn.xml",
  "postproc_model": "/path/to/pointPillars/pretrained/pointpillars_ov_postproc.xml"

File: docs/user-guide/elements/g3dinference.md (hunk 7 above)
Comment (ZiningLi) [reply docs/user-guide/elements/g3dinference.md:32]: Added warning block in doc and added GST_WARNING in code.

File: src/monolithic/gst/3d_elements/g3dinference/g3dinference.cpp (hunk 6 above)
Comment (ZiningLi) [reply src/monolithic/gst/3d_elements/g3dinference/g3dinference.cpp]: Updated. I changed score-threshold to use 0.0 as the default, with 0.0 meaning no filtering, and removed the previous -1 sentinel behavior. I also updated the related code, element documentation, and sample documentation/scripts to keep the behavior and user-facing description consistent.

File: src/monolithic/gst/elements/gvametaconvert/jsonconverter.cpp (hunk 8 above)
Comment (ZiningLi) [reply src/monolithic/gst/elements/gvametaconvert/jsonconverter.cpp:539]: Added a dedicated test for the pointpillars_3d branch and regression checks for the existing tensor/detection output format in gvametaconvert unit tests. Verified by running test_metaconvert_plugin_converting in the container, which passes.

image

File: tests/unit_tests/tests_gstgva/test_pipeline_g3dinference.py (hunk 9 above)
Comment (ZiningLi) [reply tests/unit_tests/tests_gstgva/test_pipeline_g3dinference.py:184]: Addressed.

Added tests for:
- empty point cloud input
- mismatched payload size vs. lidar_point_count
- missing config property

Validated with:
- 4/4 tests passed in test_pipeline_g3dinference.py
image

File: src/monolithic/gst/3d_elements/g3dinference/g3dinference.cpp [hunk 10]
```diff
@@ -0,0 +1,542 @@
...
+    }
+}
+
+static GstCaps *gst_g3d_inference_transform_caps(GstBaseTransform *trans, GstPadDirection direction, GstCaps *caps,
+                                                 GstCaps *filter_caps) {
+    (void)trans;
+    (void)caps;
+    (void)filter_caps;
+
+    if (direction == GST_PAD_SINK)
```
Comment (oonyshch) [inline src/monolithic/gst/3d_elements/g3dinference/g3dinference.cpp]: both branches return the same caps, so the direction check is effectively a no-op here. since sink and src both use `application/x-lidar`, this is functionally correct, but you could simplify by removing the conditional or adding a comment to clarify the passthrough is intentional

File: src/monolithic/gst/3d_elements/g3dinference/g3dinference.cpp [hunk 11]
```diff
@@ -0,0 +1,542 @@
...
+LidarMeta *get_lidar_meta(GstBuffer *buffer) {
+    return reinterpret_cast<LidarMeta *>(gst_buffer_get_meta(buffer, LIDAR_META_API_TYPE));
+}
+
+PointPillarsRuntime *get_runtime(GstG3DInference *filter) {
+    return reinterpret_cast<PointPillarsRuntime *>(filter->runtime);
+}
+
+class GstBufferMapGuard {
+  public:
```
Comment (oonyshch) [inline src/monolithic/gst/3d_elements/g3dinference/g3dinference.cpp]: consider adding deleted copy/move constructors and assignment operators to prevent accidental double-unmap (e.g., `GstBufferMapGuard(const GstBufferMapGuard&) = delete;`). not an issue with the current local usage, but would make the raii wrapper safer

**Review by oonyshch — DISMISSED** [review]
overall LGTM. please consider all comments below with suggestions for code clarity

File: src/monolithic/gst/3d_elements/g3dinference/g3dinference.cpp [hunk 12]
```diff
@@ -0,0 +1,542 @@
...
+        }
+
+        const std::filesystem::path config_dir = std::filesystem::path(config_path).parent_path();
+        auto resolve_path = [&config_dir](const std::string &path_str) {
+            const std::filesystem::path path(path_str);
+            if (path.is_absolute())
+                return path.lexically_normal().string();
+            return (config_dir / path).lexically_normal().string();
+        };
+
```
Comment (tbujewsk) [inline src/monolithic/gst/3d_elements/g3dinference/g3dinference.cpp:100]: Insert short info (samples/gstreamer/gst_launch/g3dinference/README.md) where all these components must be taken from

File: src/monolithic/gst/3d_elements/g3dinference/g3dinference.cpp [hunk 13]
```diff
@@ -0,0 +1,542 @@
...
+    GstG3DInference *filter = GST_G3D_INFERENCE(trans);
+    delete get_runtime(filter);
+    filter->runtime = NULL;
+    filter->initialized = FALSE;
+    return TRUE;
+}
+
+static GstFlowReturn gst_g3d_inference_transform_ip(GstBaseTransform *trans, GstBuffer *buffer) {
+    GstG3DInference *filter = GST_G3D_INFERENCE(trans);
+    g_mutex_lock(&filter->mutex);
```
Comment (mholowni) [inline src/monolithic/gst/3d_elements/g3dinference/g3dinference.cpp]: Small comment - consider using `GMutexLockGuard` - custom lock guard for gmutex:
[gmutex_lock_guard.h](https://github.com/open-edge-platform/dlstreamer/blob/main/src/monolithic/gst/common/gmutex_lock_guard.h)

**Review by mholowni — DISMISSED** [review]

File: src/monolithic/gst/3d_elements/g3dinference/g3dinference.cpp [hunk 14]
```diff
@@ -0,0 +1,542 @@
...
+
+LidarMeta *get_lidar_meta(GstBuffer *buffer) {
+    return reinterpret_cast<LidarMeta *>(gst_buffer_get_meta(buffer, LIDAR_META_API_TYPE));
+}
+
+PointPillarsRuntime *get_runtime(GstG3DInference *filter) {
+    return reinterpret_cast<PointPillarsRuntime *>(filter->runtime);
+}
+
+class GstBufferMapGuard {
```
Comment (mholowni) [inline src/monolithic/gst/3d_elements/g3dinference/g3dinference.cpp]: Consider moving it to another file, it could be reused by different components

File: src/monolithic/gst/3d_elements/g3dinference/g3dinference.cpp [hunk 15]
```diff
@@ -0,0 +1,542 @@
...
+  public:
+    GstBufferMapGuard(GstBuffer *buffer, GstMapInfo *map_info) : _buffer(buffer), _map_info(map_info) {
```
Comment (mholowni) [inline src/monolithic/gst/3d_elements/g3dinference/g3dinference.cpp]: Consider adding `explicit` to the constructor

File: src/monolithic/gst/3d_elements/g3dinference/g3dinference.cpp (hunk 12 above)
Comment (ZiningLi) [reply src/monolithic/gst/3d_elements/g3dinference/g3dinference.cpp:100]: Added a short note to README.md clarifying where these components come from, including the openvino_contrib repository and the specific PointPillars paths.

File: src/monolithic/gst/3d_elements/g3dinference/g3dinference.cpp (hunk 11 above)
Comment (ZiningLi) [reply src/monolithic/gst/3d_elements/g3dinference/g3dinference.cpp]: Done. Added deleted copy/move constructors and assignment operators.

File: src/monolithic/gst/3d_elements/g3dinference/g3dinference.cpp (hunk 14 above)
Comment (ZiningLi) [reply src/monolithic/gst/3d_elements/g3dinference/g3dinference.cpp]: Done. Moved GstBufferMapGuard to include/dlstreamer/gst/buffer_map_guard.h so it can be reused by other GStreamer components.

File: src/monolithic/gst/3d_elements/g3dinference/g3dinference.cpp (hunk 15 above)
Comment (ZiningLi) [reply src/monolithic/gst/3d_elements/g3dinference/g3dinference.cpp]: Done. Constructor is now marked explicit.

File: src/monolithic/gst/3d_elements/g3dinference/g3dinference.cpp (hunk 13 above)
Comment (ZiningLi) [reply src/monolithic/gst/3d_elements/g3dinference/g3dinference.cpp]: Done. Reused the existing GMutexLockGuard from src/monolithic/gst/common/gmutex_lock_guard.h. Added the common include path to the g3dinference CMakeLists.txt.

File: src/monolithic/gst/3d_elements/g3dinference/g3dinference.cpp (hunk 10 above)
Comment (ZiningLi) [reply src/monolithic/gst/3d_elements/g3dinference/g3dinference.cpp]: Good point. Simplified — removed the redundant direction check since both branches returned the same caps.

File: docs/user-guide/elements/g3dinference.md [hunk 16]
```diff
@@ -0,0 +1,185 @@
...
+```
+
+### Inference with JSON export
+
+```bash
+gst-launch-1.0 multifilesrc location="lidar/%06d.bin" caps=application/octet-stream ! \
+  g3dlidarparse ! \
+  g3dinference config=pointpillars_ov_config.json device=GPU score-threshold=0.5 ! \
+  gvametaconvert add-tensor-data=true format=json json-indent=2 ! \  
+  gvametapublish file-format=2 file-path=pointpillar.json ! \
```
Comment (oonyshch) [inline docs/user-guide/elements/g3dinference.md]: the output filename here is `pointpillar.json` (without the `s`). every other reference in the pr uses `pointpillars`

File: samples/download_public_models.sh [hunk 17]
```diff
@@ -1304,3 +1305,40 @@ if array_contains "mars-small128" "${MODELS_TO_PROCESS[@]}" || array_contains "a
     echo_color "\nModel already exists: $MODEL_DIR.\n" "yellow"
   fi
 fi
+
+
```
Comment (oonyshch) [inline samples/download_public_models.sh]: there are two blank lines here instead of one

File: src/monolithic/gst/elements/gvametapublish/base/gvametapublishbase.cpp [hunk 18]
```diff
@@ -131,6 +132,11 @@ static void gva_meta_publish_base_before_transform(GstBaseTransform *trans, GstB
     GvaMetaPublishBase *self = GVA_META_PUBLISH_BASE(trans);
 
     GST_DEBUG_OBJECT(self, "before transform");
+    if (gst_buffer_get_meta(buffer, LIDAR_META_API_TYPE) && trans->segment.format != GST_FORMAT_TIME) {
```
Comment (oonyshch) [inline src/monolithic/gst/elements/gvametapublish/base/gvametapublishbase.cpp:137]: it would help to add a brief comment explaining the intent here , bc without one it's not obvious why a lidar buffer with a time-format segment is expected to fall through to the same timestamp sync path as video:

```cpp
// lidar pipelines typically run with a default (bytes/units) segment, not a TIME segment.
// in that case the video timestamp sync below does not apply, so return early.
// if the segment format is TIME, fall through to the standard sync path.
```

File: src/monolithic/gst/3d_elements/g3dinference/g3dinference.h [hunk 19]
```diff
@@ -0,0 +1,45 @@
+/*******************************************************************************
+ * Copyright (C) 2026 Intel Corporation
+ *
+ * SPDX-License-Identifier: MIT
+ ******************************************************************************/
+
+#ifndef __GST_G3D_INFERENCE_H__
```
Comment (oonyshch) [inline src/monolithic/gst/3d_elements/g3dinference/g3dinference.h]: `buffer_map_guard.h` (also added in this pr) uses `#pragma once`. since both are new files, aligning them to the same style would keep things consistent within the pr.

File: src/monolithic/gst/elements/gvametaconvert/jsonconverter.cpp [hunk 20]
```diff
@@ -499,6 +501,114 @@ json convert_radar_process_meta(GstGvaMetaConvert *converter, GstBuffer *buffer)
     return result;
 }
 
+json get_lidar_frame_data(GstGvaMetaConvert *converter, GstBuffer *buffer, LidarMeta *lidar_meta) {
+    assert(converter && buffer && lidar_meta && "Expected valid pointers for LiDAR conversion");
```
Comment (oonyshch) [inline src/monolithic/gst/elements/gvametaconvert/jsonconverter.cpp]: `assert()` could be stripped in builds, so this check won't guard the function. consider replacing with explicit null guards and a `GST_ERROR_OBJECT` + early return, which is consistent with how the rest of the file handles unexpected inputs:

```cpp
if (!converter || !buffer || !lidar_meta) {
    GST_ERROR_OBJECT(converter, "unexpected null pointer in get_lidar_frame_data");
    return json::object();
}
```

File: src/monolithic/gst/elements/gvametaconvert/jsonconverter.cpp [hunk 21]
```diff
@@ -499,6 +501,114 @@ json convert_radar_process_meta(GstGvaMetaConvert *converter, GstBuffer *buffer)
...
+    if (dims.size() != 2 || dims[1] != 9)
+        return objects;
+
+    const size_t detection_count = dims[0];
+    const size_t detection_width = dims[1];
+    if (detection_count == 0 || !tensor.has_field("data_buffer"))
+        return objects;
+
+    const std::vector<float> data = tensor.data<float>();
+    if (data.size() < detection_count * detection_width)
```
Comment (oonyshch) [inline src/monolithic/gst/elements/gvametaconvert/jsonconverter.cpp]: if the tensor data buffer is smaller than expected, this returns an empty detections array with no debug output. adding a `GST_WARNING` here would make it easier to catch model output shape mismatches during development:

```cpp
if (data.size() < detection_count * detection_width) {
    GST_WARNING("pointpillars tensor data size (%zu) is smaller than expected (%zu x %zu). returning empty detections.",
                data.size(), detection_count, detection_width);
    return objects;
}
```

File: tests/unit_tests/tests_gstgva/test_pipeline_g3dinference.py [hunk 22]
```diff
@@ -0,0 +1,560 @@
...
+
+Gst.init(None)
+
+SCRIPT_DIR = os.path.dirname(os.path.realpath(__file__))
+DLSTREAMER_ROOT = os.path.abspath(os.path.join(SCRIPT_DIR, "..", "..", ".."))
+OPENVINO_CONTRIB_URL = "https://github.com/openvinotoolkit/openvino_contrib.git"
+# Keep the test offline-friendly by requiring pre-provisioned PointPillars assets by default.
+# Set G3DINFERENCE_ALLOW_DOWNLOAD=1 only for explicit local validation that permits fetching sources.
+ALLOW_POINTPILLARS_DOWNLOAD = os.environ.get("G3DINFERENCE_ALLOW_DOWNLOAD", "").lower() in ("1", "true", "yes")
+EXPECTED_DETECTIONS = [
```
Comment (oonyshch) [inline tests/unit_tests/tests_gstgva/test_pipeline_g3dinference.py:33]: these expected values are tied to a specific pretrained model version. if the upstream pointpillars checkpoint in `openvino_contrib` is ever updated, these will break without an obvious explanation. a short comment noting which model revision these values correspond to (e.g. a commit hash or tag) would make future maintenance easier.

File: docs/user-guide/elements/g3dinference.md (hunk 16 above)
Comment (ZiningLi) [reply docs/user-guide/elements/g3dinference.md]: Fixed, thanks. Updated it to pointpillars.json for consistency with the other references.

File: samples/download_public_models.sh (hunk 17 above)
Comment (ZiningLi) [reply samples/download_public_models.sh]: Good catch. I cleaned that up and left a single blank line there.

File: src/monolithic/gst/elements/gvametapublish/base/gvametapublishbase.cpp (hunk 18 above)
Comment (ZiningLi) [reply src/monolithic/gst/elements/gvametapublish/base/gvametapublishbase.cpp:137]: Added a brief comment to clarify the intent here: LiDAR buffers only return early for non-TIME segments, while TIME-segment LiDAR buffers fall through to the standard timestamp sync path.

File: src/monolithic/gst/3d_elements/g3dinference/g3dinference.h (hunk 19 above)
Comment (ZiningLi) [reply src/monolithic/gst/3d_elements/g3dinference/g3dinference.h]: Good point. I switched the new header to #pragma once so it matches buffer_map_guard.h and keeps the new files consistent within the PR.

File: src/monolithic/gst/elements/gvametaconvert/jsonconverter.cpp (hunk 21 above)
Comment (ZiningLi) [reply src/monolithic/gst/elements/gvametaconvert/jsonconverter.cpp]: Good point. I added a warning here so tensor size or output shape mismatches show up in the logs before returning an empty detections array.

File: tests/unit_tests/tests_gstgva/test_pipeline_g3dinference.py (hunk 22 above)
Comment (ZiningLi) [reply tests/unit_tests/tests_gstgva/test_pipeline_g3dinference.py:33]: Good point. I documented the openvino_contrib revision these expected values were captured from, and I also pinned the test-time download path to that same revision so the auto-fetched PointPillars assets stay aligned with the assertions.

File: src/monolithic/gst/elements/gvametaconvert/jsonconverter.cpp (hunk 20 above)
Comment (ZiningLi) [reply src/monolithic/gst/elements/gvametaconvert/jsonconverter.cpp]: Good point. I replaced the assert with explicit runtime null checks and error logging, so this path is still guarded in release builds instead of relying on assert being enabled.

**Review by oonyshch — APPROVED** [review]
👍LGTM

**Review by yunowo — APPROVED** [review]
